stop logging killing connection/stream because serving request timed out and response had been started

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**: it stops persisting the following error in the server's log:
```
E0921 07:09:22.814963       1 runtime.go:78] Observed a panic: &errors.errorString{s:"killing connection/stream because serving request timed out and response had been started"} (killing connection/stream because serving request timed out and response had been started)
goroutine 1088 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2039b80, 0xc000552230)
	k8s.io/apimachinery@v0.19.0-rc.2/pkg/util/runtime/runtime.go:74 +0xa6
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0xc00257dc98, 0x1, 0x1)
	k8s.io/apimachinery@v0.19.0-rc.2/pkg/util/runtime/runtime.go:48 +0x89
panic(0x2039b80, 0xc000552230)
	runtime/panic.go:969 +0x175
k8s.io/apiserver/pkg/server/filters.(*baseTimeoutWriter).timeout(0xc0026077c0, 0xc002444460)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/filters/timeout.go:257 +0x1bc
k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP(0xc0002df9a0, 0x26b7640, 0xc0005a7c00, 0xc00243ff00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/filters/timeout.go:141 +0x2f3
k8s.io/apiserver/pkg/server/filters.WithWaitGroup.func1(0x26b7640, 0xc0005a7c00, 0xc00243fe00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/filters/waitgroup.go:59 +0x137
net/http.HandlerFunc.ServeHTTP(0xc00127d650, 0x26b7640, 0xc0005a7c00, 0xc00243fe00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1(0x26b7640, 0xc0005a7c00, 0xc00243fd00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/endpoints/filters/requestinfo.go:39 +0x269
net/http.HandlerFunc.ServeHTTP(0xc00127d680, 0x26b7640, 0xc0005a7c00, 0xc00243fd00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/endpoints/filters.WithAuditAnnotations.func1(0x26b7640, 0xc0005a7c00, 0xc00243fc00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/endpoints/filters/audit_annotations.go:37 +0x142
net/http.HandlerFunc.ServeHTTP(0xc0002df9c0, 0x26b7640, 0xc0005a7c00, 0xc00243fc00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/endpoints/filters.WithWarningRecorder.func1(0x26b7640, 0xc0005a7c00, 0xc00243fb00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/endpoints/filters/warning.go:35 +0x1a7
net/http.HandlerFunc.ServeHTTP(0xc0002df9e0, 0x26b7640, 0xc0005a7c00, 0xc00243fb00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/endpoints/filters.WithCacheControl.func1(0x26b7640, 0xc0005a7c00, 0xc00243fb00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/endpoints/filters/cachecontrol.go:31 +0xa8
net/http.HandlerFunc.ServeHTTP(0xc0002dfa00, 0x26b7640, 0xc0005a7c00, 0xc00243fb00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/server/httplog.WithLogging.func1(0x26aa080, 0xc002394110, 0xc00243fa00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/httplog/httplog.go:91 +0x2f2
net/http.HandlerFunc.ServeHTTP(0xc0002dfa20, 0x26aa080, 0xc002394110, 0xc00243fa00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/server/filters.withPanicRecovery.func1(0x26aa080, 0xc002394110, 0xc00243fa00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/filters/wrap.go:51 +0xe6
net/http.HandlerFunc.ServeHTTP(0xc0002dfa40, 0x26aa080, 0xc002394110, 0xc00243fa00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/server.(*APIServerHandler).ServeHTTP(0xc00127d710, 0x26aa080, 0xc002394110, 0xc00243fa00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/handler.go:189 +0x51
net/http.serverHandler.ServeHTTP(0xc002255c00, 0x26aa080, 0xc002394110, 0xc00243fa00)
	net/http/server.go:2843 +0xa3
net/http.initALPNRequest.ServeHTTP(0x26bd780, 0xc001dc88d0, 0xc00163b500, 0xc002255c00, 0x26aa080, 0xc002394110, 0xc00243fa00)
	net/http/server.go:3415 +0x8d
golang.org/x/net/http2.(*serverConn).runHandler(0xc001cd1080, 0xc002394110, 0xc00243fa00, 0xc0026076e0)
	golang.org/x/net@v0.0.0-20200602114024-627f9648deb9/http2/server.go:2147 +0x8b
created by golang.org/x/net/http2.(*serverConn).processHeaders
	golang.org/x/net@v0.0.0-20200602114024-627f9648deb9/http2/server.go:1881 +0x505
E0921 07:09:22.815015       1 wrap.go:39] apiserver panic'd on GET /openapi/v2
```

From now on the timeout handler will panic with `http.ErrAbortHandler` when a response to the client has been already sent and the timeout elapsed. As a result, the connection will be forcefully closed `EOF` and we will record `requestAbortsTotal`.

Previously the stack trace was persisted in the logs which caused confusion and false alarms.

Additionally, a new metric `requestAbortsTotal` was defined to count aborted requests. The new metric allows for aggregation for each `group`, `version`, `verb`, `resource`, `subresource` and `scope`.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**: Initially I wanted to use `errors.Is and errors.As` but that would require to have an additional function anyway as we need to cast `err interface{}` first. Other than that that would require coupling  `errConnKilled` with `http.ErrAbortHandler` which I didn't like.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
A new metric `requestAbortsTotal` has been introduced that counts aborted requests for each `group`, `version`, `verb`, `resource`, `subresource` and `scope`.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

p0lyn0mial commented:
/assign @sttts @mfojtik

mfojtik commented:
/lgtm

sttts commented on staging/src/k8s.io/apimachinery/pkg/util/net/http_test.go:
```diff
@@ -401,8 +401,8 @@ func TestConnectWithRedirects(t *testing.T) {
 			netdialer := &net.Dialer{
-				Timeout:   wait.ForeverTestTimeout,
-				KeepAlive: wait.ForeverTestTimeout,
+				Timeout:   time.Second * 30,
```
why this? Seems to be unrelated.

sttts commented on staging/src/k8s.io/apimachinery/pkg/util/net/interface.go:
```diff
@@ -455,3 +456,23 @@ func ChooseBindAddressForInterface(intfName string) (net.IP, error) {
 	return nil, fmt.Errorf("unable to select an IP from %s network interface", intfName)
 }
+
+// AbortHandlerErrors holds a list of errors that signal the API server to abort and HTTP handler.
```
English parse error

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:
```diff
@@ -28,13 +28,21 @@ import (
 var errConnKilled = fmt.Errorf("killing connection/stream because serving request timed out and response had been started")
 
+func init() {
+	knet.AbortHandlerErrors = append(knet.AbortHandlerErrors, errConnKilled)
```
this is ugly. Writing some global state....

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:
```diff
@@ -21,14 +21,15 @@ import (
 func WithPanicRecovery(handler http.Handler) http.Handler {
 	return withPanicRecovery(handler, func(w http.ResponseWriter, req *http.Request, err interface{}) {
-		if err == http.ErrAbortHandler {
+		if net.IsAbortHandlerError(err) {
```
maybe we should at least log why we dropped the request?

sttts replied:
```
if errors.Is(http.ErrAbortHandler) {
  err := recover()
  klog.Error(err)
  panic(http.ErrAbortHandler)
}
```

p0lyn0mial replied:
from golang docs `To abort a handler so the client sees an interrupted response but the server doesn't log an error, panic with the value ErrAbortHandler`

maybe we should actually throw `ErrAbortHandler` instead of `errConnKilled` ?
that would be seen as `  timeout_test.go:220: Get "http://127.0.0.1:61805": EOF` on the client side.

p0lyn0mial committed:
stops puting a stacktrace of aborted requests in the logs

Aborted requests are the ones that were disrupted with http.ErrAbortHandler.
For example, the timeout handler will panic with http.ErrAbortHandler when a response to the client has been already sent
and the timeout elapsed.

Additionally, a new metric requestAbortsTotal was defined to count aborted requests. The new metric allows for aggregation for each group, version, verb, resource, subresource and scope.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:
```diff
@@ -29,11 +29,19 @@ import (
+			// an error, panic with the value ErrAbortHandler.
+			//
+			//
+			panic(http.ErrAbortHandler)
```
panicking with `http.ErrAbortHandler` (well any panic would do) will actually close the underlying connection.

Previously `errConnKilled` was captured, logged and the connection was left intact since calling `WriteHeader` multiple times is not supported by the std library.

I think that before closing the connection (before panicking) we could actually try to send a message to the client.

p0lyn0mial replied:
> I think that before closing the connection (before panicking) we could actually try to send a message to the client.

Even if we did, the msg would be ignored by the client (golang) library `On error, any Response can be ignored`
https://golang.org/src/net/http/client.go?s=20422:20474#L575

sttts replied:
Where did we catch errConnKilled ?

fedebongio commented:
/assign @lavalamp

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:
```diff
@@ -29,11 +29,19 @@ import (
+			// an error, panic with the value ErrAbortHandler.
+			//
+			// We are throwing http.ErrAbortHandler deliberately so that a client is notified
+			klog.Warningf("Aborting the request since ErrAbortHandler was received (%v) on %v %v. This might result in a EOF error on the client side", err, req.Method, req.RequestURI)
```
I'd prefer a metric, or a rate limit on this log message like that in https://github.com/kubernetes/kubernetes/pull/88600/files

sttts replied:
+1 for metric.

What do we have now for errors like EOF or tls errors with `--v=2`? They are also visible, aren't they? A warning here is definitely too heavy. We won't warn about EOFs either, will we?

sttts replied:
What will we see in `--v=2` for timeouts in general?

lavalamp commented:
/approve

I think my preference order is: metric, rate-limited log message, no log message, log message

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
+	}
+}
+
+var tsCrt = []byte(`-----BEGIN CERTIFICATE-----
```
will be valid for 10y

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
+	}
+	expectedOutput := "Aborting the request since ErrAbortHandler was received (net/http: abort Handler) on GET /. This might result in a EOF error on the client side"
+	defer cleanUp()
+	os.Stderr = writer
```
isn't there also a buf writer without the go routine blow?

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
+		t.Fatal("expected to receive an error")
+	}
+
+	cleanUp()
```
the defer of cleanUp does not harm?

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
+		t.Fatalf("unable to read the captured std output, err %v", err)
+	}
+	capturedOutput := string(data)
+	// maxize is the size of the expected output + 35%
```
maxSize

sttts replied:
why 35% ?

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
+	// make another req to the server
+	// the connection should be reused
+	// the client uses a custom transport that checks and panics when the con wasn't reused.
+	_, err = client.Do(req)
```
can one reuse a request?

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
 	}
 }
+
+type fakeTransport struct {
```
panicOnNonReuseTransport

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -185,6 +184,14 @@ var (
 		[]string{"verb", "group", "version", "resource", "subresource", "scope", "component", "code"},
 	)
+	// RequestAbortsTotal is a number of aborted requests with http.ErrAbortHandler
+	RequestAbortsTotal = compbasemetrics.NewCounter(
```
we need to define some labels on it because aborted request won't end up in the audit logs, see `TestErrConnKilledWithAuditFilter` test.

would it be okay to have `verb, group, version and the resource` ?

sttts replied:
yeah, would add the same we have for other request metrics.

p0lyn0mial commented:
/retest

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -206,6 +206,16 @@ var (
+	requestAbortsTotal = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_request_aborts_total",
+			Help:           "Number of requests which apiserver aborted possibly due to a timeout for each group, version, verb, resource, subresource and scope",
```
nit: due to a timeout<comma>, for each group ...

Otherwise it reads that it timed out for every group.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:
```diff
@@ -246,15 +244,17 @@ func (tw *baseTimeoutWriter) timeout(err *apierrors.StatusError) {
+		// the client sees an interrupted response but the server doesn't log
+		// an error, panic with the value ErrAbortHandler.
+		//
+		// We are throwing http.ErrAbortHandler deliberately so that a client is notified
```
...., but we want to supress a not helpful stacktrace in the logs.

p0lyn0mial commented:
/test pull-kubernetes-bazel-test

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:
```diff
@@ -17,22 +17,37 @@ limitations under the License.
+				metrics.RecordRequestAbort(req, nil)
+				return
+			}
+			metrics.RecordRequestAbort(req, info)
```
I'd like to see a rate limited log message about this.  We rate limit runtime.HandleError by default, so I'll add that here as a followup because @p0lyn0mial is in Europe and on vacation this week.

deads2k commented:
/lgtm
